nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/iot/AbstractAWSIoTProcessor.java

```diff
@@ -0,0 +1,230 @@
+     */
+    @Override
+    protected AWSIotClient createClient(final ProcessContext context, final AWSCredentials credentials, final ClientConfiguration config) {
+        getLogger().info("Creating client using aws credentials ");
+        // Actually this client is not needed. it is initialized due to the pattern of
+        // AbstractAWSProcessor
+        return new AWSIotClient(credentials, config);
+    }
+
+    /**
+     * Gets ready an MQTT client by connecting to a AWS IoT WebSocket endpoint specific to the properties
+     * @param context processor context
+     */
+    void init(final ProcessContext context) {
+        // read out properties
+        awsEndpoint = context.getProperty(PROP_ENDPOINT).getValue();
+        awsRegion = context.getProperty(REGION).getValue();
+        awsClientId = context.getProperty(PROP_CLIENT).isSet() ? context.getProperty(PROP_CLIENT).getValue() : PROP_DEFAULT_CLIENT;
+        awsKeepAliveSeconds = context.getProperty(PROP_KEEPALIVE).isSet() ? context.getProperty(PROP_KEEPALIVE).asInteger() : PROP_DEFAULT_KEEPALIVE;
+        awsTopic = context.getProperty(PROP_TOPIC).getValue();
+        awsQos = context.getProperty(PROP_QOS).isSet() ? context.getProperty(PROP_QOS).asInteger() : DEFAULT_QOS;
+        // initialize and connect to mqtt endpoint
+        mqttClient = connect(context);
+    }
+
+    @OnStopped
+    public void onStopped(final ProcessContext context) {
+        try {
+            mqttClient.disconnect();
+        } catch (MqttException me) {
+            getLogger().warn("MQTT " + me.getMessage());
+        }
+        getLogger().info("Disconnected");
+    }
+
+    /**
+     * Returns the lifetime-seconds of the established websocket-connection
+     * @return seconds
+     */
+    long getConnectionDuration() {
+        return dtLastConnect != null
+                ? TimeUnit.MILLISECONDS.toSeconds(new Date().getTime() - dtLastConnect.getTime())
+                : awsKeepAliveSeconds + 1;
+    }
+
+    /**
+     * In seconds get the remaining lifetime of the connection. It is not the actual time to
+     * expiration but an advice to when it is worth renewing the connection.
+     * @return seconds
+     */
+    long getRemainingConnectionLifetime() {
+        return awsKeepAliveSeconds - DEFAULT_CONNECTION_RENEWAL_BEFORE_KEEP_ALIVE_EXPIRATION;
+    }
+
+    /**
+     * Indicates if WebSocket connection is about to expire. It gives the caller an advice
+     * to renew the connection some time before the actual expiration.
+     * @return Indication (if true caller should renew the connection)
+     */
+    boolean isConnectionAboutToExpire() {
+        return getConnectionDuration() > getRemainingConnectionLifetime();
+    }
+
+    /**
+     * Connects to the websocket-endpoint over an MQTT client.
+     * @param context processcontext
+     * @return websocket connection client
+     */
+    MqttWebSocketAsyncClient connect(ProcessContext context) {
```

Comment:
The beginning portion of this method re-creates multiple different local variables to use in reconnecting. Maybe persist anything can be re-used (like the connections options).